session.go

```diff
@@ -338,63 +338,85 @@ func sqlForLog(sql string) string {
 	return sql
 }
 
+var sysSessionPool = sync.Pool{}
+
 // ExecRestrictedSQL implements RestrictedSQLExecutor interface.
 // This is used for executing some restricted sql statements, usually executed during a normal statement execution.
 // Unlike normal Exec, it doesn't reset statement status, doesn't commit or rollback the current transaction
 // and doesn't write binlog.
-func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (ast.RecordSet, error) {
-	s.prepareTxnCtx()
-	charset, collation := s.sessionVars.GetCharsetInfo()
-	rawStmts, err := s.ParseSQL(sql, charset, collation)
+func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rows []*ast.Row, fields []*ast.ResultField, err error) {
+	// Use special session to execute the sql.
+	var se *session
+	tmp := sysSessionPool.Get()
+	if tmp != nil {
+		se = tmp.(*session)
+	} else {
+		se, err = createSession(s.store)
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+		varsutil.SetSessionSystemVar(se.sessionVars, variable.AutocommitVar, types.NewStringDatum("1"))
+		se.sessionVars.CommonGlobalLoaded = true
+		se.sessionVars.InRestrictedSQL = true
+	}
+	defer sysSessionPool.Put(se)
+
+	var recordSets []ast.RecordSet
+	recordSets, err = se.Execute(sql)
 	if err != nil {
-		return nil, errors.Trace(err)
+		err = errors.Trace(err)
+		return
 	}
-	if len(rawStmts) != 1 {
-		log.Errorf("ExecRestrictedSQL only executes one statement. Too many/few statement in %s", sql)
-		return nil, errors.New("wrong number of statement")
+
+	for i, rs := range recordSets {
```

Comment:
The caller should satisfy the restrictions.
No check here, because `se.Execute(sql)` can't control only one sql statement run, such as this input:
`set a =3; select xxx;`